File: fsharp/AoC2021/Day02/Day02/Instruction.fs
module Instruction

type Instruction =
    | Forward of int
    | Down of int
    | Up of int

let parse (line : string) =
    match line.IndexOf ' ' with
    | -1 -> failwith $"Invalid input line: {line}"
    | p -> match line[..p - 1] with
           | "forward" -> Forward (int line[p..])
           | "down" -> Down (int line[p..])
           | "up" -> Up (int line[p..])
           | instruction -> failwith $"Invalid input line: '{line}' (instruction: '{instruction}')"

File: csharp/AoC2021/Day02/Day02/Instruction.cs
namespace Day02;

public readonly struct Instruction
{
    public readonly Direction Direction;
    public readonly int Distance;

    private Instruction(Direction direction, int distance)
    {
        Direction = direction;
        Distance = distance;
    }

    public static Instruction Parse(string s)
    {
        var p = s.IndexOf(' ');
        if (p == -1)
            throw new ArgumentException($"Invalid input: {s}.");
        var distance = int.Parse(s[(p + 1)..]);
        return s[..p] switch
        {
            "forward" => new Instruction(Direction.Forward, distance),
            "down" => new Instruction(Direction.Down, distance),
            "up" => new Instruction(Direction.Up, distance),
            _ => throw new ArgumentException($"Invalid input: {s}."),
        };
    }
}
